std/arith.asm

```diff
@@ -17,18 +23,174 @@ machine Arith(CLK32_31, operation_id){
+    let sub = |x, y| ff::sub(x, y, secp_modulus);
+    let mul = |x, y| ff::mul(x, y, secp_modulus);
+    let div = |x, y| ff::div(x, y, secp_modulus);
+
+    pol commit x1[16], y1[16], x2[16], y2[16], x3[16], y3[16];
+
+    // Selects the ith limb of x (little endian)
+    // Note that the most significant limb can be up to 32 bits; all others are 16 bits.
+    let select_limb = |x, i| fe((x >> (i * 16)) & if i < 15 { 0xffff } else { 0xffffffff });
+
+    let s_for_eq1 = |x1, y1, x2, y2| div(sub(y2, y1), sub(x2, x1));
+    let s_for_eq2 = |x1, y1| div(mul(3, mul(x1, x1)), mul(2, y1));
+
+    // Adding secp_modulus to make sure that that all numbers are positive when % is applied to it.
+    let compute_x3_int = |x1, x2, s| (s * s - x1 - x2 + 2 * secp_modulus) % secp_modulus;
+    let compute_y3_int = |x1, y1, x3, s| (s * ((x1 - x3) + secp_modulus) - y1 + secp_modulus) % secp_modulus;
+
+    // Compute quotients for the various equation.
+    // Note that we add 2**258 to it, to move it from the (-2**258, 2**258) to the (0, 2**259) range, so it can
+    // be represented as an unsigned 272-bit integer.
+    // See the comment for `product_with_p` below.
+    let compute_q0_for_eq1 = |x1, y1, x2, y2, s| (-(s * x2 - s * x1 - y2 + y1) / secp_modulus + (1 << 258));
+    let compute_q0_for_eq2 = |x1, y1, s| (-(2 * s * y1 - 3 * x1 * x1) / secp_modulus + (1 << 258));
+    let compute_q1 = |x1, x2, x3, s| (-(s * s - x1 - x2 - x3) / secp_modulus + (1 << 258));
+    let compute_q2 = |x1, y1, x3, y3, s| (-(s * x1 - s * x3 - y1 - y3) / secp_modulus + (1 << 258));
+ 
+    let limbs_to_int: expr[] -> int = |limbs| array::sum(array::map_enumerated(limbs, |i, limb| int(eval(limb)) << (i * 16)));
+
+    let x1_int: -> int = || limbs_to_int(x1);
+    let y1_int: -> int = || limbs_to_int(y1);
+    let x2_int: -> int = || limbs_to_int(x2);
+    let y2_int: -> int = || limbs_to_int(y2);
+    let x3_int: -> int = || limbs_to_int(x3);
+    let s_int: -> int = || limbs_to_int(s);
+
+    let eq1_active = || eval(selEq[1]) == 1;
+    let get_operation = || match eval(operation_id) {
+        1 => "affine_256",
+        10 => "ec_add",
+        12 => "ec_double",
+        _ => panic("Unknown operation")
+    };
+    let is_ec_operation: -> int = || match get_operation() {
+        "affine_256" => 0,
+        "ec_add" => 1,
+        "ec_double" => 1,
+    };
+
+    let s_hint: -> int = || match get_operation() {
+        "affine_256" => 0,
+        "ec_add" => s_for_eq1(x1_int(), y1_int(), x2_int(), y2_int()),
+        "ec_double" => s_for_eq2(x1_int(), y1_int()),
+    };
+
+    let q0_hint: -> int = || match get_operation() {
+        "affine_256" => 0,
+        "ec_add" => compute_q0_for_eq1(x1_int(), y1_int(), x2_int(), y2_int(), s_int()),
+        "ec_double" => compute_q0_for_eq2(x1_int(), y1_int(), s_int()),
+    };
+
+    let q1_hint: -> int = || if is_ec_operation() == 1 {
+        // TODO: Make this more readable once we have let statements
```

Comment:
Is this for performance reasons?

Reply:
Yes! So we don't compute `x1` and `x2` twice.